Add create new overlay button to Navigation block

getdave opened the pull request:
## What

Adds the ability to create new Navigation Overlays directly from the Navigation block sidebar "Overlay" panel. Users can click "Create new?" in the SelectControl help text to automatically create a new overlay template part with a unique sequential name (e.g., "Overlay", "Overlay 2", "Overlay 3").

This PR addresses the overlay creation requirement from #73080, complementing the existing overlay selection functionality.

## Why

This improves the workflow for creating navigation overlays by removing the need to navigate away from the Navigation block to create overlay template parts. It matches the pattern established in PR #73389 and provides a seamless inline creation experience, fulfilling part of the requirements outlined in #73080.

## How

- Added "Create new?" button inline within the SelectControl help text using `createInterpolateElement`
- Implemented automatic unique name generation using sequential naming (Overlay, Overlay 2, etc.)
- Uses store state (`isSavingEntityRecord`) for tracking save operations instead of local state
- Automatically selects the newly created overlay and optionally navigates to it for editing
- Added template part utility functions (`parseTemplatePartId`, `getUniqueTemplatePartTitle`, `getCleanTemplatePartSlug`) to `utils.js` with reference comments to original implementations
- Includes proper error handling and loading states

## ⚠️ Out of scope

- Placement and styling of the `Edit` button. This can be handled in followups.
- Whether the "Create" action should trigger a modal to allow users to choose a title for the Overlay. 
- The precise design implementation of this UI. We can follow up on this.

## Screenshots

<img width="367" height="193" alt="Screen Shot 2025-12-16 at 12 43 14" src="https://github.com/user-attachments/assets/60dd1ff1-4076-4e2f-a1b8-891931bd99f8" />



## Testing Instructions

1. Open the Navigation block sidebar
2. Navigate to the "Overlay" panel
3. Click "Create new?" in the help text below the overlay selector
4. Verify a new overlay is created with a unique name
5. Verify the new overlay is automatically selected
6. Verify the overlay appears in the dropdown list
7. Test creating multiple overlays to verify sequential naming works correctly
8. Test error handling by attempting to create when offline or with invalid permissions

getdave committed:
Add create new overlay button to Navigation block

- Add 'Create new overlay' button to OverlayTemplatePartSelector component
- Automatically generates unique sequential names (Overlay, Overlay 2, etc.)
- Uses store state (isSavingEntityRecord) for tracking save state
- Auto-selects newly created overlay and optionally navigates to it
- Includes proper error handling and loading states

getdave committed:
Move utils

getdave committed:
Refactor overlay creation to use createInterpolateElement in help text

- Move 'Create new overlay' button into SelectControl help text using createInterpolateElement
- Change button text to 'Create new?' to match prototype
- Add template part utilities (parseTemplatePartId, getUniqueTemplatePartTitle, getCleanTemplatePartSlug) to utils.js
- Copy utilities from fields package with reference comments
- Use useCallback for handleCreateOverlay to optimize memoization
- Fix hooks order by moving useMemo before early return
- Add __next40pxDefaultSize prop to Button in help text

getdave commented:
As it stands this implementation is too complex. Looking to simplify

getdave committed:
Simplify saving state

getdave committed:
Remove passing content to create empty

getdave committed:
Avoid manually constructing the template id when we have it from the creation

getdave committed:
Decouple creation and update handling

getdave committed:
Avoid custom loading UI

getdave committed:
Remove reundant invalidation

getdave committed:
remove redundant util

getdave committed:
Update tests

getdave committed:
Tweak test wording

getdave committed:
Slim down tests

getdave committed:
More tweaks to test names

getdave committed:
Add test coverage for create

getdave committed:
Test hook

getdave committed:
Add more tests coverage

getdave commented on packages/block-library/src/navigation/edit/utils.js:
```diff
@@ -110,3 +111,47 @@ export function getNavigationChildBlockProps( innerBlocksColors ) {
+ * @param {Object} templateParts The array of template part entities.
+ * @return {string} A unique template part title.
+ */
+export const getUniqueTemplatePartTitle = ( title, templateParts ) => {
+	const lowercaseTitle = title.toLowerCase();
+	const existingTitles = templateParts.map( ( templatePart ) =>
+		templatePart.title.rendered.toLowerCase()
+	);
+
+	if ( ! existingTitles.includes( lowercaseTitle ) ) {
+		return title;
+	}
+
+	let suffix = 2;
+	while ( existingTitles.includes( `${ lowercaseTitle } ${ suffix }` ) ) {
+		suffix++;
+	}
+
+	return `${ title } ${ suffix }`;
+};
+
+/**
+ * Get a valid slug for a template part.
+ * Currently template parts only allow latin chars.
+ * The fallback slug will receive suffix by default.
+ *
+ * This implementation is copied from:
+ * packages/fields/src/components/create-template-part-modal/utils.js
+ *
+ * @param {string} title The template part title.
+ * @return {string} A valid template part slug.
+ */
+export const getCleanTemplatePartSlug = ( title ) => {
+	return kebabCase( title ).replace( /[^\w-]+/g, '' ) || 'wp-custom-part';
+};
```
Added some tests

getdave commented on packages/block-library/src/navigation/edit/overlay-template-part-selector.js:
```diff
@@ -135,20 +125,86 @@ export default function OverlayTemplatePartSelector( {
+		} catch ( error ) {
+			const errorMessage =
+				error instanceof Error &&
+				'code' in error &&
+				error.message &&
+				error.code !== 'unknown_error'
```
Comment to explain the pattern

jeryj commented on packages/block-library/src/navigation/edit/overlay-template-part-selector.js:
```diff
@@ -135,20 +125,90 @@ export default function OverlayTemplatePartSelector( {
+				className="wp-block-navigation__overlay-create-link"
+			>
+				{ __( 'Create new?' ) }
+			</Button>
```
"Create new?" doesn't provide context about what will happen. I think "Create New Overlay Template" as a separate button will be clearer and provide more confidence for ATs, also it would simplify as you don't need to use all the `createInterpolateElement` pieces. 

Especially if we're not focused on the exact UX for now but building the required tools, then having a simpler technical implementation would aid us in moving faster.

getdave replied:
I've revised the design to follow the pattern suggest by Design in https://github.com/WordPress/gutenberg/pull/73957#issuecomment-3646158247.

<img width="367" height="193" alt="Screen Shot 2025-12-16 at 12 43 14" src="https://github.com/user-attachments/assets/e75935d6-8eb6-4973-aaa4-cee1f85a5f5e" />


This addresses your concerns but may add new concerns (we'll see)!

Let's run with it and see how it feels. If it's not working we can update the visuals but I'd really like to land an implementation of "create" if possible 🙏

jeryj commented on packages/block-library/src/navigation/edit/overlay-template-part-selector.js:
```diff
@@ -135,20 +125,90 @@ export default function OverlayTemplatePartSelector( {
 		} );
 	};
 
+	const handleCreateOverlay = useCallback( async () => {
```
I think returning the typical `isResolving` could be useful within the `useCreateOverlayTemplatePart` hook. Having to implement custom isCreating and error handling for each place we want to be able to create overlays could be more tedious.  What do you think about bundling more things within `useCreateOverlayTemplatePart` instaed?

getdave replied:
I kept `isCreating` local because it's UI-specific to this component instance. If the hook is used in multiple places, each instance would have its _own_ state, which can be confusing and mismatched to expectations.

The long-term solution is to track creation via core-data resolver status, which would provide shared, operation-level state that all consumers can use. However this is challenging because then you have to work out the id of the template part _before_ it's created in order to track the creartion status via resolvers 😓 

For now, keeping it local is appropriate since it's tied to this specific button's UI state.

jeryj commented on packages/block-library/src/navigation/edit/overlay-template-part-selector.js:
```diff
@@ -161,11 +221,7 @@ export default function OverlayTemplatePartSelector( {
-						? __( 'No overlays found.' )
-						: __( 'Select an overlay to use for the navigation.' )
-				}
+				help={ helpText }
```
It's odd to have Create new? within the help text.

getdave replied:
Resolved by https://github.com/WordPress/gutenberg/pull/73919#discussion_r2623138732

jeryj reviewed (commented):
<img width="298" height="126" alt="Overlay Template form with select dropdown saying Default (none) and help text No Overlays Found. Create new?" src="https://github.com/user-attachments/assets/7cfc4f44-eef8-46f3-882d-fd84633f166a" />

Seems odd to have a dropdown saying None (default) with no other options, and then provide a link to create a new one. My quick thoughts:

- If there are no custom overlay options, remove the select field and just offer the explanation and create new overlay template button.
- Be clearer about how None does not mean NO Overlay, but No CUSTOM Overlay template. At the moment, it would be valid for someone to think they don't have ANY overlay as opposed to no Overlay templates. We'll sort this out later though.

### Bugs

#### Create empty overlay template then go back
Steps:
1. Create new?
2. Save the empty template
3. Click < Back from the Command Center
4. Bug: You'll see the new page Select pattern modal

#### Creating a second template without reloading opens an empty template editor (no canvas visible)
1. From the site editor, go through the Create new? flow above
2. Click Create new? again
3. The template editor will load without any canvas visible until you click

https://github.com/user-attachments/assets/228758d7-4bde-4c81-8bf7-4df7d614a19d

Probably out of scope, but want to document it here in case you think it's connected:

#### Deleting overlay template doesn't sever connection
If I delete the current selected overlay template, I'll still get presented with the Edit button when I go to the navigation block overlay template selection

getdave committed:
Fix Edit button visibility when overlay template is deleted

When a selected overlay template part is deleted, the Edit button was still visible even though the template part no longer exists. This fix ensures the Edit button only appears when:
- overlay attribute has a value
- template parts have been resolved (hasResolved)
- the selected template part exists in the resolved list

This prevents the Edit button from appearing when the overlay selector correctly shows 'None (default)' after deletion.

getdave committed:
Improve accessibility of create overlay button

- Change button text from 'Create new?' to 'Create New Overlay Template' for better clarity
- Add aria-label='Create new overlay template' for screen reader support
- Update tests to use the new accessible button name

getdave committed:
Refactor create overlay button to icon-based button with tooltip

- Move create button from help text to top-right position using absolute positioning
- Change from text-based button to icon-only button with plus icon
- Add tooltip using label prop with showTooltip
- Remove createInterpolateElement usage and simplify help text
- Add CSS positioning styles using SCSS variables for alignment

jeryj commented on packages/block-library/src/navigation/edit/overlay-template-part-selector.js:
```diff
@@ -135,23 +121,77 @@ export default function OverlayTemplatePartSelector( {
+				disabled={ isCreateButtonDisabled }
+				accessibleWhenDisabled
+				isBusy={ isCreating }
+				label={ __( 'Create new overlay template' ) }
```
This doesn't work with the `.show-icon-labels` option, but that seems like a core component issue, not a blocking issue in this PR. I'll address that as a follow-up root fix.

jeryj commented on packages/block-library/src/navigation/editor.scss:
```diff
@@ -654,3 +654,16 @@ body.editor-styles-wrapper .wp-block-navigation__responsive-container.is-menu-op
 	width: fit-content;
 }
+
+.wp-block-navigation__overlay-selector {
+	position: relative;
+}
+
+.wp-block-navigation__overlay-create-button {
+	position: absolute;
+	// Align with SelectControl label by offsetting by button padding
+	// Small icon buttons have 0 padding, but we align with default button top padding (4px = $grid-unit-05)
+	top: -$grid-unit-05;
+	right: 0;
+	z-index: 1;
+}
```
Nit: I think this could be addressed with <Flex> or <HStack> components like the Styles -> Colors -> Edit Palette use of the + button.

jeryj reviewed (approved):
Let's bring this in and iterate. I like this approach better than the previous.
